### ISSUE TITLE
Enhancement: `only-throw-error`: allow throwing `never`

### ISSUE THREAD

**Issue by AlCalzone** [typescript-eslint/typescript-eslint#10054]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/only-throw-error/

### Description

Although TypeScript should understand that `never`-returning functions never return, but there are some cases where this does not work. A workaround is to throw the result of the function call (which is `never`).
However when this is done, the `only-throw-error` rule complains about not throwing an error.

### Fail

```TypeScript
The existing examples are good enough. This issue is about adding an exception to the rule.
```

### Pass

```TypeScript
function neverReturns(): never {
  throw new Error();
}

function fail(): never {
  throw neverReturns();
  // ^ this should be okay, currently an error.
  // throwing may be necessary to satisfy the TS compiler
}
```

**Comment by auvred** [comment]
> but there are some cases where this does not work.

Could you please provide some examples of this? Looking at the above code, I don't see the need to allow `never` typed values to be thrown.

**Comment by AlCalzone** [comment]
```ts
const neverReturns = ()=>{
	throw new Error();
}

function fail(): never {
	throw neverReturns();
}
```

would be one example. If you remove the `throw` in `fail()`, TS complains. If you don't, ESLint complains.

FWIW, this can be avoided by annotating `neverReturns` with `: () =>never`. I originally had a more [complex example](https://github.com/microsoft/TypeScript/issues/60059), but as it turns out, that was also a matter of annotating the types explicitly.

**Comment by JoshuaKGoldberg** [comment]
> FWIW, this can be avoided by annotating neverReturns with : () =>never

Yup, that's the way it's meant to work. TypeScript doesn't automatically infer a `never` return type. `neverReturns` is `() => void` as-is. Since Typescript sees it as _"some function with a `void` return"_, that's what lint rules see it as too.

The right fix here isn't to adjust linting, it's to properly inform TypeScript as noted:

```diff
- const neverReturns = () => {
+ const neverReturns = (): never =>{
	throw new Error();
}

function fail(): never {
-	throw neverReturns();
+	neverReturns();
}
```

Are there any examples that don't boil down to _"manually add `: never`"_?

**Comment by kirkwaiblinger** [comment]
Huh, I find TS's behavior strange here tbh. For me, hovering in the editor over `neverReturns` does show `() => never`, regardless which of the following you do:
```ts
const neverReturns = () => { throw new Error() }
const neverReturns = (): never => { throw new Error() }
const neverReturns: () => never = () => { throw new Error() }`.
```

But only `const neverReturns: () => never` silences the TS error. So I agree with the conclusion that the linter is correct here, but I can see why this question came up, since it seems like TS is misleadingly reporting the wrong type to the user(?)

**Comment by bradzacher** [comment]
I agree that a motivating case would be good, and without one I don't think we can accept this.

We currently (begrudgingly) have options for any/unknown because you can get either type from a catch variable, and it is useful to be able to re-throw the variable without first asserting its type.

But I currently can't think of a reason why you'd explicitly throw a `never` - I'd expect that any "exhaustive logic" `never` would be first wrapped in an error eg `` throw new Error(`Unhandled case ${val}`); `` and so you wouldn't directly throw it.

And the usecase of re-throwing a never returning function is moot cos you can just call it (as mentioned above).

**Comment by AlCalzone** [comment]
Fair enough. I had accepted throwing `never` as a workaround to force TS to accept that a method call won't return, and updating typescript-eslint broke that workaround.

Only after this issue I learned that there's a difference for control flow between
```ts
interface ObjectWithNeverReturningMethod {
  fail(): never;
}

const foo = bar as ObjectWithNeverReturningMethod;
foo.fail();
```
and
```ts
const foo: ObjectWithNeverReturningMethod = bar as any;
foo.fail();
```
even though the type of `foo` and `foo.fail()` looks like it is the same in both cases.

I'm okay with calling this a TypeScript oddity and closing this feature request.

**Comment by kirkwaiblinger** [comment]
> I'm okay with calling this a TypeScript oddity and closing this feature request.

For a related situation that I recently wrestled with, it is _surprisingly difficult_ to call a conditionally determined assertion function:
```ts
declare function assert1(x: unknown): asserts x;
declare function assert2(x: unknown): asserts x;

const randomAssert = Math.random() > 0.5 ? assert1 : assert2;

randomAssert(true) // not allowed!

// solution:
const randomAssert: typeof assert1 | typeof assert2 = Math.random() > 0.5 ? assert1 : assert2;
randomAssert(true) // allowed 
```

all very odd! But agreed let's close this out. Thanks for discussing!

**Comment by Josh-Cena** [comment]
I agree that allowing `never` is a bad idea. Usually `never` comes from exhaustive control flow, so if you ever reach there, the variable is essentially `unknown` because there's nothing that can be proven about it. Directly throwing defeats the purpose of the rule, which is to prevent non-error values _you_ created to be thrown. Rethrowing an `unknown` is fine since the origin is unknown and is not worse than not catching at all.
